source/v1/I2CDetail.cpp

```diff
@@ -0,0 +1,293 @@
+}
+
+void I2CResourceManager::process_event(uint32_t event) {
+    I2CTransaction * t;
+    CORE_UTIL_ASSERT(TransactionQueue != nullptr);
+    if (!TransactionQueue) {
+        return;
+    }
+    // Get the current item in the transaction queue
+    t = TransactionQueue;
+    if (!t) {
+        return;
+    }
+    // If the event is 0, no further action is required
+    if (!event) {
+        return;
+    }
+    // Fire the irqcallback for the segment
+    t->call_irq_cb(event);
+    // This can't be done with atomics: there are too many side-effects.
+    mbed::util::CriticalSectionLock lock;
+    // If there is another segment to process, advance the segment pointer
+    // Record whether there was another segment
+    bool TransactionDone = !t->advance_segment();
+    // If there was an event that is not a complete event
+    // or there was a complete event and the next segment is nullptr
+    if ((event & I2C_EVENT_ALL & ~I2C_EVENT_TRANSFER_COMPLETE) ||
+        ((event & I2C_EVENT_TRANSFER_COMPLETE) && TransactionDone))
+    {
+        // fire the handler
+        minar::Scheduler::postCallback(I2C_event_callback_t(this, &I2CResourceManager::handle_event).bind(t,event));
+        // Advance to the next transaction
+        TransactionQueue = t->get_next();
+        if (TransactionQueue) {
+            // Initiate the next transaction
+            start_transaction();
+        } else {
+        }
+    } else if (!TransactionDone) {
+        start_segment();
+    }
+}
+
+void I2CResourceManager::handle_event(I2CTransaction *t, uint32_t event) {
+    t->process_event(event);
+    // This happens after the callbacks have all been called
+    t->get_issuer()->free(t);
+}
+
+I2CResourceManager::I2CResourceManager() : TransactionQueue(nullptr) {}
+
+I2CResourceManager::~I2CResourceManager() {
+    mbed::util::CriticalSectionLock lock;
+    while (TransactionQueue) {
+        I2CTransaction * tx = (TransactionQueue);
+        TransactionQueue = tx->get_next();
+        tx->get_issuer()->free(tx);
+    }
+}
+
+class HWI2CResourceManager : public I2CResourceManager {
+public:
+    HWI2CResourceManager(const size_t id, void(*handler)(void)):
+        _i2c(),
+        _id(id),
+        _usage(DMA_USAGE_NEVER),
+        _inited(false),
+        _handler(handler)
+    {}
+
+    I2CError init(PinName sda, PinName scl) {
+        // calling init during a transaction could cause communication artifacts
+        if (!_inited) {
+            // The init function also set the frequency to 100000
+            i2c_init(&_i2c, sda, scl);
+            _inited = true;
+        } else {
+            CORE_UTIL_ASSERT_MSG(_scl == scl && _sda == sda, "Each I2C peripheral may only be used on one set of pins");
+            // Only support an I2C master on a single pair of pins
+            if (_scl != scl || _sda != sda) {
+                return I2CError::PinMismatch;
+            }
+        }
+        return I2CError::None;
+    }
+
+    I2CError start_segment () {
```

Comment:
Brace placement on member functions such as this does not conform to the style guide.